lab/README.md

```diff
@@ -0,0 +1,1086 @@
+dp-simple-kmod-deviceplugin-e383247e62b56585-qn2fb       1/1     Running     0          48m
+dp-simple-kmod-driver-build-e383247e62b56585-1-build     0/1     Completed   0          50m
+dp-simple-kmod-driver-container-e383247e62b56585-wrktm   1/1     Running     0          48m
+```
+
+
+And in our very specific example we could check if our “dummy” devices are present in the driver-container:
+
+
+```
+[root@ebelarte chart]# oc exec -it dp-simple-kmod-driver-container-e383247e62b56585-wrktm -- printenv | grep DUMMY
+DUMMY_DEVICES=dev_1
+```
+
+
+
+
+## EXERCISE 5
+
+
+Troubleshoot failure to load a kmod
+There are some different places we could look for logs and traces regarding the load or build of our modules with SRO.
+
+Different examples:
+
+- Driver container not being created. No pod running. If we inspect the cluster events we could see that our serviceaccount is not allowed to use some of the settings in the deployment.
+
+
+```
+oc get events
+90m         Warning   FailedCreate       daemonset/dp-simple-kmod-driver-container-e383247e62b56585   Error creating: pods "dp-simple-kmod-driver-container-e383247e62b56585-" is forbidden: unable to validate against any security context constraint: [provider "anyuid": Forbidden: not usable by user or serviceaccount, provider restricted: .spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used, spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used, spec.containers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed, spec.containers[0].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used, provider "nonroot": Forbidden: not usable by user or serviceaccount, provider "ootmodprobe": Forbidden: not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount, provider "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount, provider "privileged": Forbidden: not usable by user or serviceaccount]
+```
+
+
+
+
+* BuildConfig not running. We could inspect the logs in the special-resource-controller pod (manager container) and look for possible issues:
+
+
+2022-06-20T16:04:40.461Z	INFO	dp-simple-kmod  	RECONCILE REQUEUE: Could not reconcile chart	{"error": "cannot reconcile hardware states: failed to create state templates/0000-buildconfig.yaml: after CRUD hooks failed: could not wait for resource: Waiting too long for resource: timed out waiting for the condition "}
+
+…
+2022-06-20T16:06:55.286Z	INFO	warning  	OnError: node Conflict Label specialresource.openshift.io/state-dp-simple-kmod-0000 err %!s(<nil>)
+
+…
```

Comment:
````suggestion
```
2022-06-20T16:04:40.461Z	INFO	dp-simple-kmod  	RECONCILE REQUEUE: Could not reconcile chart	{"error": "cannot reconcile hardware states: failed to create state templates/0000-buildconfig.yaml: after CRUD hooks failed: could not wait for resource: Waiting too long for resource: timed out waiting for the condition "}
…
2022-06-20T16:06:55.286Z	INFO	warning  	OnError: node Conflict Label specialresource.openshift.io/state-dp-simple-kmod-0000 err %!s(<nil>)
…
```
````